lib/razor/cli/query.rb

```diff
@@ -51,12 +56,40 @@ def run
     while @segments.any?
       nav = @segments.shift
       @parse.stripped_args << nav
+
+      # If the endpoint exposes a depth parameter, then that means
+      # we are navigating to a collection endpoint. If this is our
+      # final endpoint (i.e. we exit out of the while loop after making
+      # our request), then setting 'depth' to 1 will return the full
+      # details of each item in the collection in a single request, so we
+      # do not need to make N additional requests to each item's endpoint.
+      # If, however, this is an intermediate endpoint, which occurs when
+      # we have a nested query (e.g. something like `razor nodes <node_name>`),
+      # then setting 'depth' to 0 will return a list of references. This is OK
+      # and optimal, because we will be querying only a single entry in that list.
+      #
+      # We do not expose the 'depth' param to the user, because it controls how
+      # we traverse the collection and that is something handled by the razor-client.
+      #
+      # Note: We delete our knowledge of the depth parameter after each request to avoid
+      # having it carry-over to subsequent endpoints.
+      @has_depth_param = false
+      @options.delete('depth')
       @segments = get_optparse(@doc, nav).order(@segments)
+      if @has_depth_param
+        # @segments.any? is how we detect an intermediate collection endpoint
+        @options['depth'] = @segments.any? ? 0 : 1
+      end
+
       @doc = @navigate.move_to nav, @doc, @options
     end
 
     # Get the next level if it's a list of objects.
     if @doc.is_a?(Hash) and @doc['items'].is_a?(Array)
+      # If our nav endpoint had a depth parameter, then we will already have
+      # the next level.
+      return @doc if @options['depth']
```

Comment:
It'd be great to turn `@options` into a local variable that only exists in the `while` loop above, but I can't figure a clean way to handle this check.

Reply:
There are some instance variables that are not needed in the Query class (eg @options, @doc, @queryoptparse, etc.). I chose not to touch/clean those up in this PR, b/c that'll break consistency with some of the other classes (there are other classes that also probably have unnecessary instance variables). As mentioned on HipChat, that clean-up effort would probably be more suitable if we ever decide to rewrite parts of the `razor-client`. It is not something that belongs in this PR.